indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java

```diff
@@ -0,0 +1,995 @@
+    );
+  }
+
+  public static Path getJobPath(JobID jobID, Path workingDirectory)
+  {
+    return new Path(workingDirectory, jobID.toString());
+  }
+
+  public static Path getTaskPath(JobID jobID, TaskAttemptID taskAttemptID, Path workingDirectory)
+  {
+    return new Path(getJobPath(jobID, workingDirectory), taskAttemptID.toString());
+  }
+
+  public static void cleanup(Job job) throws IOException
+  {
+    final Path jobDir = getJobPath(job.getJobID(), job.getWorkingDirectory());
+    final FileSystem fs = jobDir.getFileSystem(job.getConfiguration());
+    fs.delete(jobDir, true);
+  }
+
+
+  public static HadoopDruidConverterConfig converterConfigFromConfiguration(Configuration configuration)
+      throws IOException
+  {
+    final String property = Preconditions.checkNotNull(
+        configuration.get(HadoopDruidConverterConfig.CONFIG_PROPERTY),
+        HadoopDruidConverterConfig.CONFIG_PROPERTY
+    );
+    return HadoopDruidConverterConfig.fromString(property);
+  }
+
+  public static void converterConfigIntoConfiguration(
+      HadoopDruidConverterConfig priorConfig,
+      List<DataSegment> segments,
+      Configuration configuration
+  )
+  {
+    final HadoopDruidConverterConfig config = new HadoopDruidConverterConfig(
+        priorConfig.getDataSource(),
+        priorConfig.getInterval(),
+        priorConfig.getIndexSpec(),
+        segments,
+        priorConfig.isValidate(),
+        priorConfig.getDistributedSuccessCache(),
+        priorConfig.getHadoopProperties(),
+        priorConfig.getJobPriority(),
+        priorConfig.getSegmentOutputPath()
+    );
+    try {
+      configuration.set(
+          HadoopDruidConverterConfig.CONFIG_PROPERTY,
+          HadoopDruidConverterConfig.jsonMapper.writeValueAsString(config)
+      );
+    }
+    catch (JsonProcessingException e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  private final HadoopDruidConverterConfig converterConfig;
+  private long loadedBytes = 0L;
+  private long writtenBytes = 0L;
+
+  public HadoopConverterJob(
+      HadoopDruidConverterConfig converterConfig
+  )
+  {
+    this.converterConfig = converterConfig;
+  }
+
+  public List<DataSegment> run() throws IOException
+  {
+    final JobConf jobConf = new JobConf();
+    jobConf.setKeepFailedTaskFiles(false);
+    for (Map.Entry<String, String> entry : converterConfig.getHadoopProperties().entrySet()) {
+      jobConf.set(entry.getKey(), entry.getValue(), "converterConfig.getHadoopProperties()");
+    }
+    final List<DataSegment> segments = converterConfig.getSegments();
+    if (segments.isEmpty()) {
+      throw new IAE(
+          "No segments found for datasource [%s]",
+          converterConfig.getDataSource()
+      );
+    }
+    converterConfigIntoConfiguration(converterConfig, segments, jobConf);
+    jobConf.setNumMapTasks(segments.size());
+    jobConf.setNumReduceTasks(0);
+    jobConf.setWorkingDirectory(new Path(converterConfig.getDistributedSuccessCache()));
```

Comment:
Because working directory is a first class configuration object, and the distributed success cache is used as a working directory.